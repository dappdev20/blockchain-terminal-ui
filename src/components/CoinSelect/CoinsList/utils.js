import { sortBy } from 'lodash';

const isSearched = (item, query, filter) => {
  const symbol = item.symbol || '';
  const name = item.name || '';
  const lowerCaseQuery = query.toLowerCase();

  let filterFlag = true;
  if (filter !== 'All') {
    if (filter === 'Fiat') {
      filterFlag = symbol.includes('F:');
    } else if (filter === 'Stocks') {
      filterFlag = symbol.includes('S:');
    } else if (filter === 'Crypto') {
      filterFlag = !symbol.includes('F:') && !symbol.includes('S:');
    }
  }

  let weight = 9999;
  if (query) {
    const index1 = symbol
      .replace('S:', '')
      .replace('F:', '')
      .toLowerCase()
      .indexOf(lowerCaseQuery);
    if (index1 !== -1) {
      weight = index1;
    } else {
      const index2 = name.toLowerCase().indexOf(lowerCaseQuery);
      if (index2 !== -1) {
        weight = index2 + 100;
      }
    }

    if (!filterFlag) {
      weight += 1000;
    }
  } else {
    if (filterFlag) {
      weight = 0;
    }
  }

  return weight;
};

export const getTableItems = ({ topGroupItems = [], mainItems = [] }, { searchInputValue, filterItem }) => {
  let tableItems = [];

  let searchedTopGroupItemsWeights = [];
  topGroupItems.forEach(item => {
    const weight = isSearched(item, searchInputValue, filterItem);
    if (weight !== 9999) {
      searchedTopGroupItemsWeights.push({
        weight,
        item
      });
    }
  });

  searchedTopGroupItemsWeights = sortBy(searchedTopGroupItemsWeights, item => item.weight);
  tableItems = searchedTopGroupItemsWeights.map(val => val.item);

  let searchedMainItemsWeights = [];
  mainItems.forEach(item => {
    const weight = isSearched(item, searchInputValue, filterItem);
    if (weight !== 9999) {
      if (!topGroupItems.find(({ symbol }) => symbol === item.symbol)) {
        searchedMainItemsWeights.push({
          weight,
          item
        });
      }
    }
  });

  searchedMainItemsWeights = sortBy(searchedMainItemsWeights, item => item.weight);
  tableItems = tableItems.concat(searchedMainItemsWeights.map(val => val.item));

  return tableItems;
};
