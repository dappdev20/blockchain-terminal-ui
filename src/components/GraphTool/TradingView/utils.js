import { INTERVAL_MAP } from './historyProvider';

export const getSymbol = ({ exchange, market }) => ({
  name: market.code,
  ticker: `${exchange.code}:${market.code}`,
  description: market.code,
  timezone: 'UTC',
  type: 'market',
  session: '24x7',
  minmov: 1,
  minmov2: 0,
  pricescale: 1e2,
  pointvalue: 1,
  volume_precision: 4,
  data_status: 'streaming',
  has_intraday: true,
  has_daily: true,
  supported_resolutions: ['1', '5', '15', '60', '360', '1440', 'D'],
  intraday_multipliers: ['1', '5', '15', '60', '360', '1440', 'D'],
  has_weekly_and_monthly: false,
  has_empty_bars: false,
  exchange_name: exchange.code,
  exchange: exchange.code,
  market_name: market.code
});

const roundTimeByInterval = (time, interval) => +new Date(Math.trunc(time / interval) * interval);

export const drawHistoryGraph = (
  bars,
  firstDataRequest,
  onDataCallback,
  symbolInfo,
  resolution,
  onErrorCallback,
  apiDataLoadObservable
) => {
  let normalizedBars;
  const url = `https://rest.qa.bct.trade/api/best-conversion-rate/${symbolInfo.name}`;
  const barsLength = bars.length;

  fetch(url)
    .then(response => response.json())
    .then(Data => {
      let price = 0;
      let average1 = 0;
      let average2 = 0;

      try {
        price = Data.data.rate || 0;
      } catch (err) {
        price = 0;
      }

      try {
        average1 = (bars[barsLength - 1].low + bars[barsLength - 1].high) / 2;
        average2 = (1 / bars[barsLength - 1].low + 1 / bars[barsLength - 1].high) / 2;
      } catch (e) {
        average1 = 0;
        average2 = 0;
      }

      if (Math.abs(average1 - price) > Math.abs(average2 - price)) {
        normalizedBars = bars.map((b, i) => ({
          close: 1 / +b.cls,
          isLastBar: i === bars.length - 1,
          time: +new Date(b.time),
          volume: 1 / +b.vol,
          isBarClosed: true,
          open: 1 / +b.opn,
          high: 1 / +b.high,
          low: 1 / +b.low
        }));
      } else {
        normalizedBars = bars.map((b, i) => ({
          close: +b.cls,
          isLastBar: i === bars.length - 1,
          time: +new Date(b.time),
          volume: +b.vol,
          isBarClosed: true,
          open: +b.opn,
          high: +b.high,
          low: +b.low
        }));
      }

      if (firstDataRequest) {
        // TODO handle correct
        try {
          this.lastBarTime = roundTimeByInterval(
            normalizedBars[normalizedBars.length - 1].time,
            INTERVAL_MAP[resolution]
          );
        } catch (e) {
          // console.log(e);
        }
      }
      apiDataLoadObservable.next({
        apiLoaded: true
      });
      return onDataCallback(normalizedBars, { noData: !normalizedBars.length });
    })
    .catch(err => onErrorCallback());
};

export const drawHistoryGraphWithMultiply = (
  bars,
  firstDataRequest,
  onDataCallback,
  symbolInfo,
  resolution,
  onErrorCallback,
  apiDataLoadObservable
) => {
  // console.log('[symbolInfo111]', symbolInfo);
  const coin1 = symbolInfo.name.split('-')[0] || '';
  const coin2 = symbolInfo.name.split('-')[1] || '';

  let normalizedBars;
  fetch(`https://rest.qa.bct.trade/api/best-conversion-rate/${coin1}-ETH`)
    .then(response => response.json())
    .then(Data => {
      let rateY = 0;
      try {
        rateY = Data.data.rate || 0;
      } catch (err) {
        rateY = 0;
      }

      fetch(`https://rest.qa.bct.trade/api/best-conversion-rate/${coin2}-BTC`)
        .then(response => response.json())
        .then(Data => {
          let rateZ = 0;
          try {
            rateZ = Data.data.rate || 0;
          } catch (err) {
            rateZ = 0;
          }

          const rate = rateZ !== 0 ? rateY / rateZ : 0;

          normalizedBars = bars.map((b, i) => ({
            close: +b.cls * rate,
            isLastBar: i === bars.length - 1,
            time: +new Date(b.time),
            volume: +b.vol * rate,
            isBarClosed: true,
            open: +b.opn * rate,
            high: +b.high * rate,
            low: +b.low * rate
          }));

          if (firstDataRequest) {
            // TODO handle correct
            try {
              this.lastBarTime = roundTimeByInterval(
                normalizedBars[normalizedBars.length - 1].time,
                INTERVAL_MAP[resolution]
              );
            } catch (e) {
              console.log(e);
            }
          }
          apiDataLoadObservable.next({
            apiLoaded: true
          });
          return onDataCallback(normalizedBars, { noData: !normalizedBars.length });
        })
        .catch(err => onErrorCallback());
    })
    .catch(err => onErrorCallback());
};

export const customIndicatorsGetter = PineJS => {
  return Promise.resolve([
    {
      name: 'Equity',
      metainfo: {
        _metainfoVersion: 40,
        id: 'Equity@tv-basicstudies-1',
        scriptIdPart: '',
        name: 'Equity',
        description: 'Equity',
        shortDescription: 'Equity',

        is_hidden_study: true,
        is_price_study: true,
        isCustomIndicator: true,

        plots: [{ id: 'plot_0', type: 'line' }],
        defaults: {
          styles: {
            plot_0: {
              linestyle: 0,
              visible: true,

              // Make the line thinner
              linewidth: 1,

              // Plot type is Line
              plottype: 2,

              // Show price line
              trackPrice: true,

              transparency: 40,

              // Set the plotted line color to dark red
              color: '#880000'
            }
          },

          // Precision is set to one digit, e.g. 777.7
          precision: 1,

          inputs: {}
        },
        styles: {
          plot_0: {
            // Output name will be displayed in the Style window
            title: 'Equity value',
            histogramBase: 0
          }
        },
        inputs: []
      },

      constructor: function() {
        this.init = function(context, inputCallback) {
          this._context = context;
          this._input = inputCallback;

          var symbol = 'BINANCE:BTC-USDT';
          this._context.new_sym(symbol, PineJS.Std.period(this._context), PineJS.Std.period(this._context));
        };

        this.main = function(context, inputCallback) {
          this._context = context;
          this._input = inputCallback;

          this._context.select_sym(1);

          var v = PineJS.Std.close(this._context);
          return [v];
        };
      }
    }
  ]);
};
