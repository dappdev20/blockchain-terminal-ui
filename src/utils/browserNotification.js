const showNotification = (title = '', body = '', icon = '', image = '', requireInteraction = false, actionCallback) => {
  Notification.requestPermission(permission => {
    if (permission === 'granted') {
      const notification = new Notification(title, {
        icon,
        image,
        body,
        requireInteraction
      });

      notification.onclick = () => {
        if (actionCallback) {
          actionCallback();
        }
      };
    }
  });
};

export default showNotification;
