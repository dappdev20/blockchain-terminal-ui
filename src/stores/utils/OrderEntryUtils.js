import React from 'react';

import {
  ProgramId,
  ORDER_ENTRY_ROUTE,
  ORDER_TICKET_TICKET_ID,
  ORDER_TICKET_CLIENT_ID,
  ORDER_TICKET_PROGRAM_ID,
  ORDER_TICKET_SYMBOL,
  ORDER_TICKET_SIZE,
  ORDER_TICKET_PRICE,
  ORDER_TICKET_SIDE,
  ORDER_TICKET_ROUTE
} from '../../config/constants';
import { customDigitFormat } from '../../utils';

export const TicketId = (() => {
  let orderCount = 0;
  return () => `${ProgramId}-${++orderCount}`;
})();

const orderTicketDefaults = {
  [ORDER_TICKET_ROUTE]: ORDER_ENTRY_ROUTE
};

export const OrderTicketPayload = (baseSymbol, quoteSymbol, amount, price, side, clientIdForOrder, programId) => ({
  ...orderTicketDefaults,
  [ORDER_TICKET_CLIENT_ID]: clientIdForOrder,
  [ORDER_TICKET_TICKET_ID]: TicketId(),
  [ORDER_TICKET_PROGRAM_ID]: programId,
  [ORDER_TICKET_SYMBOL]: [`${baseSymbol}-${quoteSymbol}`],
  [ORDER_TICKET_SIZE]: Number(amount),
  [ORDER_TICKET_PRICE]: price,
  [ORDER_TICKET_SIDE]: side
});

const onlyFloatChars = (s = '') => s.toString().replace(/[^\d.]/g, '');
const wasPercentage = ([zero, dot]) => dot === '.' || zero === '.';
const shouldBePercentage = (s = '') => s.length >= 2 && s[0] === '0' && s[1] !== '.';
const asPercentage = ([zero, ...s]) => `${zero}.${s.join('')}`;
const only1Dot = (s = '') =>
  s
    .replace(/[.]/, '{dot}')
    .replace(/[.]/, '')
    .replace('{dot}', '.');
const cleanNumToStr = (s = '') =>
  Number(s)
    .toString()
    .trim();

export const valueNormalized = (prevInput, input = '') => {
  const inputStr = String(input);
  const prevInputStr = String(prevInput);

  const num = onlyFloatChars(inputStr);
  const shouldBePc = shouldBePercentage(num);
  const wasPercent = wasPercentage(prevInputStr);

  if (inputStr.length < 1) return inputStr;
  if (wasPercent && shouldBePc) return cleanNumToStr(only1Dot(num.slice(1)));
  return shouldBePc ? asPercentage(num) : only1Dot(num);
};

export const postSubmissionSnackbarHandler = (
  type,
  snackbar,
  { [ORDER_TICKET_SIDE]: side, [ORDER_TICKET_SIZE]: amount, [ORDER_TICKET_PRICE]: price, [ORDER_TICKET_SYMBOL]: symbol }
) => {
  let baseSymbol = '';
  if (symbol && symbol.length > 0) {
    const symbols = symbol[0].split('-');
    if (symbols && symbols.length > 0) {
      baseSymbol = symbols[0];
    }
  }

  snackbar({
    message: () => (
      <>
        <span>
          <b>{type} Order Submitted!</b>
        </span>{' '}
        <br />
        <br />
        <span>
          <b>Side:</b> {side} {baseSymbol}
        </span>{' '}
        <br />
        <span>
          <b>Amount:</b> {customDigitFormat(amount)} {baseSymbol}
        </span>{' '}
        <br />
        <span>
          <b>Price:</b> {customDigitFormat(price)} USDT
        </span>{' '}
        <br />
        <span>
          <b>Total Price:</b> {customDigitFormat(amount * price)} USDT
        </span>
      </>
    )
  });
};

export const getOrderPrice = order => order[ORDER_TICKET_PRICE];
export const getOrderAmount = order => order[ORDER_TICKET_SIZE];
export const getOrderSymbol = order => order[ORDER_TICKET_SYMBOL];
export const getOrderSide = order => order[ORDER_TICKET_SIDE];
