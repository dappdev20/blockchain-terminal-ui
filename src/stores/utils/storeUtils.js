import distanceInWords from 'date-fns/distance_in_words';
import parse from 'date-fns/parse';
import format from 'date-fns/format';
import moment from 'moment';

import { roundToFixedNum } from '@/utils';

export const isDiffArr = (a = [], b = []) => {
  let i = -1;
  const alen = a.length;

  while (++i < alen) {
    if (a[i] !== b[i]) {
      return true;
    }
  }
  return false;
};

export const transformOrderBookData = (data, noise) => {
  // allways from smaller to bigger
  const size = data.length;
  const result = [];
  let cumulativeAmount = 0;
  for (let i = 0; i < size; i++) {
    const [price, amount, exchanges] = data[i];
    cumulativeAmount += amount;

    result.push({
      price: price + noise,
      amount,
      cost: price * amount,
      exchange: exchanges ? exchanges.toUpperCase() : '',
      total: i < size - 1 ? data[i + 1][0] : price,
      cumulativeAmount
    });
  }

  return result;
};

export const processOrderBookData = data => {
  let totalAmount = 0;
  let totalCost = 0;
  let maxAmount = 0;
  let maxCost = 0;
  let maxPrice = 0;

  data.forEach(([price, amount]) => {
    const cost = price * amount;

    totalAmount += amount;
    totalCost += cost;

    maxAmount = Math.max(maxAmount, amount);
    maxPrice = Math.max(maxPrice, price);
    maxCost = Math.max(maxCost, cost);
  });

  return {
    totalAmount,
    totalCost,
    maxAmount,
    maxCost,
    maxPrice
  };
};

export const reverseOrderBookPrices = (data, spotPrice) => {
  const reversedRate = 1 / spotPrice;
  return data.map(([price, amount, exchanges]) => [
    (price / spotPrice) * reversedRate,
    amount / reversedRate,
    exchanges
  ]);
};

export const convertOrderBookPrices = (data, priceRate, amountRate) => {
  return data.map(([price, amount, exchanges]) => [price * priceRate, amount * amountRate, exchanges]);
};

export const shiftMapStoreFromArray = (mapStore = new Map(), updateArr = [], maxRows = 0) => {
  const updateArrLen = updateArr.length;

  if (updateArrLen >= maxRows) {
    let i = -1;
    while (updateArr[++i] && i < maxRows) {
      mapStore.set(i, updateArr[i]);
    }
  } else {
    const tmp = new Map();
    const carryLimit = maxRows - updateArrLen;
    let i = -1;
    let j = updateArrLen - 1;

    // cache old indices
    while (++i < carryLimit && mapStore.has(i) && ++j < maxRows) {
      tmp.set(j, mapStore.get(i));
    }

    // then prepend incoming updateArr to mapStore
    i = -1;
    while (++i < updateArrLen) {
      mapStore.set(i, updateArr[i]);
    }

    /* eslint-disable-next-line */
    for (let [k, v] of tmp) {
      mapStore.set(k, v);
    }
  }
};

export const formatNumForDisplay = num => roundToFixedNum(num, 2);

export const calculateFee = (size, price, precision) => formatNumForDisplay(size * price * 0.001, precision);

export const getTimeFormatted = time => distanceInWords(parse(time), new Date(), { includeSeconds: true });

export const scheduleVisualDOMUpdate = fn => {
  requestAnimationFrame(fn);
};

export const getDateFormatted = time => format(new Date(time), 'MM.DD.YYYY');

export const getNewDateFormatted = time => {
  return moment(time).format('MMM D');
};
