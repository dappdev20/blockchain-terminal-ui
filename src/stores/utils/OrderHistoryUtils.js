import partial from 'lodash.partial';
import get from 'lodash.get';
import {
  CREATED_ORDER_STATUS_LABEL,
  FILLED_ORDER_STATUS_LABEL,
  OrderHistoryStoreLabels,
  PARTIALLY_FILLED_ORDER_STATUS_LABEL,
  UNKNOWN_ORDER_STATUS_LABEL
} from '../../config/constants';
import { getDateFormatted, getNewDateFormatted, getTimeFormatted, formatNumForDisplay } from './storeUtils';
import { sortObjectArray, customDigitFormat } from '../../utils';
import { currencies } from '../../mock/currencies';

export const makeOrderHistoryRequest = (ClientId, ProgramId, Skip, Limit, SentTime, { SubmitOrderHistoryRequest }) => {
  SubmitOrderHistoryRequest({
    ClientId,
    ProgramId,
    SentTime,
    Skip,
    Limit
  });
};

export const getOrderHistoryRequestPartial = ({ ClientId, ProgramId, Skip, PageSize, SentTime }) =>
  partial(makeOrderHistoryRequest, ClientId, ProgramId, Skip, PageSize, SentTime);

export const updateOrderHistoryError = e => console.log(e);

export const getStatus = ({ Message = '' }) => {
  if (Message === '') {
    return UNKNOWN_ORDER_STATUS_LABEL;
  }
  if (Message === 'Filled') {
    return FILLED_ORDER_STATUS_LABEL;
  }
  if (Message === 'FilledPartially') {
    return PARTIALLY_FILLED_ORDER_STATUS_LABEL;
  }
  if (Message === 'Created') {
    return CREATED_ORDER_STATUS_LABEL; // Not sure about this
  }
  return UNKNOWN_ORDER_STATUS_LABEL;
};

// Temporary until backend returns OriginalSize consistently instead of Amount
const getOriginalAmount = ({ Amount, OriginalSize }) => {
  return Number.isNaN(Amount) ? OriginalSize : Amount;
};

// Data used for Trade History OrderTabs
export const mapTradeHistoryForDisplay = Orders =>
  Orders.map(({ OriginalSize, Amount, Price, SentTime, Symbol, Side, Exchange, TicketId }) => {
    try {
      const [Base, Quote] = Side === 'Sell' ? Symbol.split('-').reverse() : Symbol.split('-');
      const OriginalAmount = getOriginalAmount({ Amount: Number(Amount), OriginalSize: Number(OriginalSize) });
      return {
        date: `${getDateFormatted(SentTime)}`,
        timeUnFormatted: SentTime,
        amountExchanged: `${formatNumForDisplay(OriginalAmount)} ${Base}`,
        amountReceived: `${formatNumForDisplay(OriginalAmount * Price)} ${Quote}`,
        exchange: Exchange,
        ticketId: TicketId,

        // Meta Data
        amount: OriginalAmount,
        price: Price,
        base: Base,
        quote: Quote
      };
    } catch (e) {
      return {};
    }
  });

export const getFirstTradeHistoryPair = ({ base = null, quote = null }) => [base, quote];

// export const formatOrderHistoryDataForDisplay = ({Tickets}) => {
//     return Tickets.reduce((accumulator, {Orders = []}) => accumulator.concat(mapOrdersForDisplay(Orders)), [])
// };

export const formatTradeHistoryDataForDisplay = (mapOrdersForDisplay, { Tickets }) => {
  return Tickets.reduce((accumulator, curr) => accumulator.concat(mapOrdersForDisplay([curr])), []);
};

export const formatTradeHistoryData = targetTickets =>
  sortObjectArray(
    'timeUnFormatted',
    formatTradeHistoryDataForDisplay(mapTradeHistoryForDisplay, { Tickets: targetTickets })
  ); // Todo: Bug Here on earliest times

export const getTotalTradeExchangedAndReceived = Orders =>
  Orders.reduce(
    (accumulator, { amount, price } = {}) => ({
      totalExchanged: accumulator.totalExchanged + amount,
      totalReceived: accumulator.totalReceived + amount * price
    }),
    { totalExchanged: 0, totalReceived: 0 }
  );

/**
 * Function returning an object full of meta data for orders for a ticketId.
 *
 * @param {Array} tradeHistory
 *      Expects a list of Orders for one ticketId.
 *
 * @return {object} {
 *      tradeHistory: {array} original unmodified array param
 *      exchangeCount: {number} length of tradeHistory array
 *      exchangeCount: {string} length of tradeHistory array
 *      basePair: {string} base pair of first element in tradeHistory array
 *      quotePair: {string} quote pair of first element in tradeHistory array
 *      totalExchanged: {number} total exchanged in tradeHistory array
 *      totalReceived: {number} total received in tradeHistory array
 * }
 */

export const createMetaDataForTradeHistoryTargetTicketId = tradeHistory => {
  const [base, quote] = getFirstTradeHistoryPair(get(tradeHistory, '[0]', []));

  return {
    tradeHistory,
    exchangeCount: tradeHistory.length,
    basePair: base,
    quotePair: quote,
    ...getTotalTradeExchangedAndReceived(tradeHistory)
  };
};

// Data used for both Open Orders and Order History OrderTabs
export const formatOrderHistoryDataForDisplay = (Bases, Tickets, countries) => {
  return Tickets.filter(ticket => Number(ticket.Price) !== 0 && Number(ticket.ConversionAmount) !== 0).map(
    ({
      Amount,
      AmountFilled,
      ConversionAmount,
      Exchange,
      Message,
      Price,
      SentTime,
      Side,
      Size,
      Status,
      Symbol,
      TicketId,
      Type,
      UserSettings
    }) => {
      const userCurrency = countries.find(country => country.currencyCode === UserSettings.defaultFiat.toUpperCase());
      const conversionRate = userCurrency ? userCurrency.price : 1;
      const userCurrencyInfo = currencies.find(
        currency => UserSettings.defaultFiat && currency.code === UserSettings.defaultFiat.toUpperCase()
      );
      const userCurrencySymbol = userCurrencyInfo
        ? userCurrencyInfo.symbol
          ? userCurrencyInfo.symbol
          : userCurrencyInfo.code
        : '';
      try {
        const [Base, Quote] = Symbol.split('-');
        const c1Index = Bases.findIndex(c => c === Base);
        const c2Index = Bases.findIndex(c => c === Quote);
        const isAdvanced = c1Index < c2Index;
        return {
          // Displayed
          advancedMode: isAdvanced,
          filled: !isAdvanced ? AmountFilled : ConversionAmount,
          price: customDigitFormat(!isAdvanced || Price === 0 ? Price : 1 / Price, 9),
          total: !isAdvanced ? Number(ConversionAmount * conversionRate) : Number(AmountFilled * conversionRate),
          sourceTotal: !isAdvanced ? Number(ConversionAmount) : Number(AmountFilled),
          sourceFilled: !isAdvanced ? Number(AmountFilled) : Number(ConversionAmount),
          size: Size, //
          time: `${getTimeFormatted(SentTime)} ago`,
          date: `${getNewDateFormatted(SentTime)}`,
          L1: !isAdvanced ? Quote : Base,
          L2: !isAdvanced ? Base : Quote,
          type: Type,
          side: Side,
          triggers: 'xxx',
          average: 'xxx',
          [OrderHistoryStoreLabels.Ticket.Status.Label]: Status,
          // Meta Data
          exchange: Exchange,
          timeUnFormatted: SentTime,
          orderId: TicketId,
          amount: Amount,
          message: Message,
          isFailed: Number(Price) === 0 || Number(ConversionAmount) === 0,
          symbol: Symbol,
          userDefaultFiat: UserSettings.defaultFiat,
          userCurrencySymbol
        };
      } catch (e) {
        console.log(e);
        return {};
      }
    }
  );
};
