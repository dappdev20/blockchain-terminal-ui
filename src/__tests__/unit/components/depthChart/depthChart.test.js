import { partitionBuysSells, getInnerQuartileRangeOrders } from '../../../../stores/utils/storeUtils';

const AllOrders = [
    [0.05598, 38.5353, 'Sell'],
    [0.05602, 38.2899, 'Sell'],
    [0.05607, 31.6107, 'Sell'],
    [0.05613, 4.58, 'Sell'],
    [0.05616, 5.348332, 'Sell'],
    [0.05618, 48.983591, 'Sell'],
    [0.05619, 0.375525, 'Sell'],
    [0.05621, 0.509501, 'Sell'],
    [0.05622, 0.45, 'Sell'],
    [0.05623, 33.1087, 'Sell'],
    [0.05625, 15, 'Sell'],
    [0.0564, 0.725636, 'Sell'],
    [0.05641, 2.08, 'Sell'],
    [0.05644, 0.527551, 'Sell'],
    [0.05645, 3.301883, 'Sell'],
    [0.05646, 1.075877, 'Sell'],
    [0.05647, 5.355121, 'Sell'],
    [0.05648, 16.56485, 'Sell'],
    [0.05649, 137.467738, 'Sell'],
    [0.0565, 267.503401, 'Sell'],
    [0.05651, 12.183912, 'Sell'],
    [0.05652, 131.170038, 'Sell'],
    [0.05653, 27.440091, 'Sell'],
    [0.05654, 11.658855, 'Sell'],
    [0.05655, 198.709572, 'Sell'],
    [0.05656, 48.414833, 'Sell'],
    [0.05657, 337.265529, 'Sell'],
    [0.05658, 117.590932, 'Sell'],
    [0.05659, 111.579327, 'Sell'],
    [0.0566, 144.119634, 'Sell'],
    [0.05661, 166.945563, 'Sell'],
    [0.05662, 60.917347, 'Sell'],
    [0.05801, 3.200578, 'Buy'],
    [0.058, 0.519333, 'Buy'],
    [0.05793, 0.010132, 'Buy'],
    [0.05791, 0.141181, 'Buy'],
    [0.05783, 0.011197, 'Buy'],
    [0.0572, 0.004032, 'Buy'],
    [0.057, 2.316119, 'Buy'],
    [0.05689, 0.006344, 'Buy'],
    [0.05669, 0.23, 'Buy'],
    [0.05668, 0.018, 'Buy'],
    [0.05665, 0.012, 'Buy'],
    [0.05663, 0.1, 'Buy'],
    [0.05662, 0.067, 'Buy'],
    [0.05659, 0.0036, 'Buy'],
    [0.05658, 30, 'Buy'],
    [0.05656, 14.213778, 'Buy'],
    [0.05651, 6, 'Buy'],
    [0.0565, 6.5051, 'Buy'],
    [0.05649, 3.577169, 'Buy'],
    [0.05646, 2.645, 'Buy'],
    [0.05645, 0.575161, 'Buy'],
    [0.05644, 5.544292, 'Buy'],
    [0.05643, 12.69, 'Buy'],
    [0.05642, 0.106895, 'Buy'],
    [0.05641, 4.590461, 'Buy'],
    [0.0564, 52.604524, 'Buy'],
    [0.05639, 87.890628, 'Buy'],
    [0.05638, 42.867022, 'Buy'],
    [0.05637, 196.489507, 'Buy'],
    [0.05636, 116.937606, 'Buy'],
    [0.05635, 243.82495, 'Buy'],
    [0.05634, 57.615363, 'Buy'],
    [0.05633, 134.618913, 'Buy'],
    [0.05632, 571.805039, 'Buy'],
    [0.05631, 162.867853, 'Buy'],
    [0.0563, 2.692719, 'Buy'],
    [0.05625, 0.088887, 'Buy']
];

describe('test', () => {
    //sort order is tested in each new function as the wrong sort in any function can break the chart

    test('IQR function returns the same Snapshot', () => {
        const iqr = getInnerQuartileRangeOrders(AllOrders);
        expect(iqr).toMatchSnapshot();
    });

    test('Data is partitioned back into orignal order buy and sells', () => {
        const iqr = getInnerQuartileRangeOrders(AllOrders);
        const [buys, sells] = partitionBuysSells(iqr, 'Buy', 'Sell');

        expect(buys.every(buyTriplet => buyTriplet[2] === 'Buy')).toBe(true);

        buys.forEach((buyTriplet, idx) => {
            if (idx > 0) {
                expect(buyTriplet[0]).toBeLessThanOrEqual(buys[idx - 1][0]);
            }
        });

        expect(sells.every(sellTriplet => sellTriplet[2] === 'Sell')).toBe(true);

        sells.forEach((sellTriplet, idx) => {
            if (idx > 0) {
                expect(sellTriplet[0]).toBeGreaterThanOrEqual(sells[idx - 1][0]);
            }
        });
    });
});
