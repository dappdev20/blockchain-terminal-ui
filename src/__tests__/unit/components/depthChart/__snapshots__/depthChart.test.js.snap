// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test IQR function returns the same Snapshot 1`] = `
Array [
  Array [
    0.05602,
    38.2899,
    "Sell",
  ],
  Array [
    0.05607,
    31.6107,
    "Sell",
  ],
  Array [
    0.05613,
    4.58,
    "Sell",
  ],
  Array [
    0.05616,
    5.348332,
    "Sell",
  ],
  Array [
    0.05618,
    48.983591,
    "Sell",
  ],
  Array [
    0.05619,
    0.375525,
    "Sell",
  ],
  Array [
    0.05621,
    0.509501,
    "Sell",
  ],
  Array [
    0.05622,
    0.45,
    "Sell",
  ],
  Array [
    0.05623,
    33.1087,
    "Sell",
  ],
  Array [
    0.05625,
    15,
    "Sell",
  ],
  Array [
    0.0564,
    0.725636,
    "Sell",
  ],
  Array [
    0.05641,
    2.08,
    "Sell",
  ],
  Array [
    0.05644,
    0.527551,
    "Sell",
  ],
  Array [
    0.05645,
    3.301883,
    "Sell",
  ],
  Array [
    0.05646,
    1.075877,
    "Sell",
  ],
  Array [
    0.05647,
    5.355121,
    "Sell",
  ],
  Array [
    0.05648,
    16.56485,
    "Sell",
  ],
  Array [
    0.05649,
    137.467738,
    "Sell",
  ],
  Array [
    0.0565,
    267.503401,
    "Sell",
  ],
  Array [
    0.05651,
    12.183912,
    "Sell",
  ],
  Array [
    0.05652,
    131.170038,
    "Sell",
  ],
  Array [
    0.05653,
    27.440091,
    "Sell",
  ],
  Array [
    0.05654,
    11.658855,
    "Sell",
  ],
  Array [
    0.05655,
    198.709572,
    "Sell",
  ],
  Array [
    0.05656,
    48.414833,
    "Sell",
  ],
  Array [
    0.05657,
    337.265529,
    "Sell",
  ],
  Array [
    0.05658,
    117.590932,
    "Sell",
  ],
  Array [
    0.05659,
    111.579327,
    "Sell",
  ],
  Array [
    0.0566,
    144.119634,
    "Sell",
  ],
  Array [
    0.05661,
    166.945563,
    "Sell",
  ],
  Array [
    0.05662,
    60.917347,
    "Sell",
  ],
  Array [
    0.05689,
    0.006344,
    "Buy",
  ],
  Array [
    0.05669,
    0.23,
    "Buy",
  ],
  Array [
    0.05668,
    0.018,
    "Buy",
  ],
  Array [
    0.05665,
    0.012,
    "Buy",
  ],
  Array [
    0.05663,
    0.1,
    "Buy",
  ],
  Array [
    0.05662,
    0.067,
    "Buy",
  ],
  Array [
    0.05659,
    0.0036,
    "Buy",
  ],
  Array [
    0.05658,
    30,
    "Buy",
  ],
  Array [
    0.05656,
    14.213778,
    "Buy",
  ],
  Array [
    0.05651,
    6,
    "Buy",
  ],
  Array [
    0.0565,
    6.5051,
    "Buy",
  ],
  Array [
    0.05649,
    3.577169,
    "Buy",
  ],
  Array [
    0.05646,
    2.645,
    "Buy",
  ],
  Array [
    0.05645,
    0.575161,
    "Buy",
  ],
  Array [
    0.05644,
    5.544292,
    "Buy",
  ],
  Array [
    0.05643,
    12.69,
    "Buy",
  ],
  Array [
    0.05642,
    0.106895,
    "Buy",
  ],
  Array [
    0.05641,
    4.590461,
    "Buy",
  ],
  Array [
    0.0564,
    52.604524,
    "Buy",
  ],
  Array [
    0.05639,
    87.890628,
    "Buy",
  ],
  Array [
    0.05638,
    42.867022,
    "Buy",
  ],
  Array [
    0.05637,
    196.489507,
    "Buy",
  ],
  Array [
    0.05636,
    116.937606,
    "Buy",
  ],
  Array [
    0.05635,
    243.82495,
    "Buy",
  ],
  Array [
    0.05634,
    57.615363,
    "Buy",
  ],
  Array [
    0.05633,
    134.618913,
    "Buy",
  ],
  Array [
    0.05632,
    571.805039,
    "Buy",
  ],
  Array [
    0.05631,
    162.867853,
    "Buy",
  ],
  Array [
    0.0563,
    2.692719,
    "Buy",
  ],
  Array [
    0.05625,
    0.088887,
    "Buy",
  ],
]
`;
